content/articles/2019-09-tidyr-1-0-0.Rmarkdown

```diff
@@ -0,0 +1,316 @@
+  alive = purrr::map_lgl(character, "alive")
+)
+```
+
+Learn more in [`vignette("rectangle")`](http://tidyr.tidyverse.org/dev/articles/rectangle.html).
+
+## Nesting
+
+`nest()` and `unnest()` have been updated with new interfaces that are more closely aligned to evolving tidyverse conventions. The biggest change is to their operation with multiple columns: 
+
+```{R, eval = FALSE}
+# old
+df %>% nest(x, y, z)
+# new
+df %>% nest(data = c(x, y, z))
+
+# old
+df %>% unnest(x, y, z)
+# new
+df %>% unnest(c(x, y, z))
+```
+
+I've done my best to ensure that common uses of `nest()` and `unnest()` will continue to work, generating an informative warning telling you precisely how you need to update your code. Please [file an issue](https://github.com/tidyverse/tidyr/issues/new) if I've missed an important use case.
+
+Behind the scenes, I discovered that nesting (and unnesting) can be decomposed into the combination of two simpler operations:
+
+* `pack()` and `unpack()` pack and unpack multiple columns into a single data 
+  frame columns. 
+  
+* `chop()` and `unchop()` chop up rows into sets of list-columns. 
+
+<p><img src="/images/tidyr-1-0-0/nest-pack-chop.png" width="480" height="352" /></p>
+
+
+This is primarily of theoretical interest, but it considerably simplified the implementation of `nest()`, and you may occassionally find the underlying functions useful when working with exotic data structures.
+
+## `expand_grid()`
+
+[`expand_grid()`](https://tidyr.tidyverse.org/dev/reference/expand_grid.html) completes the existing family of `expand()`, `nesting()`, and `crossing()` with a low-level function that works with vectors:
+
+```{r}
+expand_grid(
+  x = 1:3,
+  y = letters[1:3],
+  z = LETTERS[1:3]
+)
+```
+
+Compared to the existing base function `expand.grid()`, `expand_grid()`:
+
+* Varies the first element most slowly (not most quickly).
+* Never converts strings to factors and doesn't add any additional attributes.
+* Returns a tibble, not a data frame.
+* Can expand data frames.
+
+The last feature is quite powerful, as it allows you to generate partial grids:
+
+```{r}
+students <- tribble(
+  ~ school, ~ student,
+  "A",      "John",
+  "A",      "Mary",
+  "A",      "Susan",
+  "B",      "John"
+)
+expand_grid(students, semester = 1:2)
+```
+
+This is made possible by the [vctrs](https://vctrs.r-lib.org) package. vctrs is primarily of interest to package developers but I wanted to talk about it briefly here because I've been having a lot of fun working with it. It's hard to concisely describe what vctrs is, but one part of it is carefully defining what a "vector" is, and provding a set of useful functions that work on all types of vctrs without any special cases. One interesting finding is that thinking of a data frame as a vector of _rows_ (not columns, as R usually does) is suprisingly useful, and something you can expect to see in more places in the tidyverse in the future. 
```

Comment:
```suggestion
This is made possible by the [vctrs](https://vctrs.r-lib.org) package. vctrs is primarily of interest to package developers but I want to talk about it briefly here because I've been having a lot of fun working with it. It's hard to concisely describe vctrs, but one aspect is carefully defining what a "vector" is, and providing a set of useful functions that work on all types of vctrs, without any special cases. One interesting finding is that thinking of a data frame as a vector of _rows_ (not columns, as R usually does) is suprisingly useful, and something you can expect to see in more places in the tidyverse in the future. 
```

Reply:
I'm really into the present tense whenever it works ⏰